presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/CrossJoinWithArrayNotContainsToAntiJoin.java

```diff
@@ -0,0 +1,224 @@
+        }
+        CallExpression arrayDistinct = call(functionAndTypeManager, "array_distinct", new ArrayType(element.getType()), array);
+        VariableReferenceExpression arrayDistinctVariable = context.getVariableAllocator().newVariable(arrayDistinct);
+        PlanNode project = PlannerUtils.addProjections(inputWithArray, context.getIdAllocator(), ImmutableMap.of(arrayDistinctVariable, arrayDistinct));
+        VariableReferenceExpression unnestVariable = context.getVariableAllocator().newVariable("field", element.getType());
+        UnnestNode unnest = new UnnestNode(inputWithArray.getSourceLocation(),
+                context.getIdAllocator().getNextId(),
+                project,
+                project.getOutputVariables(),
+                ImmutableMap.of(arrayDistinctVariable, ImmutableList.of(unnestVariable)),
+                Optional.empty());
+
+        PlanNode newLeftNode;
+
+        if (arrayAtLeftInput) {
+            newLeftNode = joinNode.getRight();
+        }
+        else {
+            newLeftNode = joinNode.getLeft();
+        }
+
+        // if element is not a VariableReferenceExpression, push the expression into a Project node so the variable can be used in equijoins
+        VariableReferenceExpression keyVariable;
+        if (element instanceof VariableReferenceExpression) {
+            keyVariable = (VariableReferenceExpression) element;
+        }
+        else {
+            keyVariable = context.getVariableAllocator().newVariable(element);
+            newLeftNode = addProjections(newLeftNode, context.getIdAllocator(), ImmutableMap.of(keyVariable, element));
+        }
+
+        JoinNode.EquiJoinClause equiJoinClause = new JoinNode.EquiJoinClause(keyVariable, unnestVariable);
+
+        List<VariableReferenceExpression> newOutputColumns = Stream.concat(newLeftNode.getOutputVariables().stream(), unnest.getOutputVariables().stream()).collect(Collectors.toList());
+
+        JoinNode newJoinNode = new JoinNode(joinNode.getSourceLocation(),
+                context.getIdAllocator().getNextId(),
+                JoinNode.Type.LEFT,
+                newLeftNode,
+                unnest,
+                ImmutableList.of(equiJoinClause),
+                newOutputColumns,
+                joinNode.getFilter(),
+                Optional.empty(),
+                Optional.empty(),
+                joinNode.getDistributionType(),
+                joinNode.getDynamicFilters());
+
+        RowExpression isNull = specialForm(IS_NULL, BOOLEAN, ImmutableList.of(unnestVariable));
+        remainingConjuncts.add(isNull);
+        FilterNode filterNode = new FilterNode(node.getSourceLocation(), context.getIdAllocator().getNextId(), newJoinNode, and(remainingConjuncts));
+        PlanNode result = restrictOutput(filterNode, context.getIdAllocator(), joinNode.getOutputVariables());
+        return Result.ofPlanNode(result);
+    }
+
+    private boolean isFromSubquery(Context context, PlanNode node)
+    {
```

Comment:
rename to isFromScalarSubquery

Reply:
Also can we move this to the pattern match?

Reply:
putting it in the pattern match is a bit messy because of disjunctions on the operator type and what side of the join this came from